File: conformance/tests/p384-primitives/p384-primitives.runar.zig
const runar = @import("runar");

pub const P384Primitives = struct {
    pub const Contract = runar.SmartContract;

    expectedPoint: runar.P384Point,

    pub fn init(expectedPoint: runar.P384Point) P384Primitives {
        return .{ .expectedPoint = expectedPoint };
    }

    pub fn verify(self: *const P384Primitives, k: i64, basePoint: runar.P384Point) void {
        const result = runar.p384Mul(basePoint, k);
        runar.assert(runar.p384OnCurve(result));
        runar.assert(result == self.expectedPoint);
    }

    pub fn verifyAdd(self: *const P384Primitives, a: runar.P384Point, b: runar.P384Point) void {
        const result = runar.p384Add(a, b);
        runar.assert(runar.p384OnCurve(result));
        runar.assert(result == self.expectedPoint);
    }

    pub fn verifyMulGen(self: *const P384Primitives, k: i64) void {
        _ = self;
        const result = runar.p384MulGen(k);
        runar.assert(runar.p384OnCurve(result));
        runar.assert(result == self.expectedPoint);
    }
};

File: conformance/tests/p384-primitives/p384-primitives.runar.go
//go:build ignore

package contract

import "runar"

type P384Primitives struct {
	runar.SmartContract
	ExpectedPoint runar.P384Point `runar:"readonly"`
}

func (c *P384Primitives) Verify(k runar.Bigint, basePoint runar.P384Point) {
	result := runar.P384Mul(basePoint, k)
	runar.Assert(runar.P384OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}

func (c *P384Primitives) VerifyAdd(a runar.P384Point, b runar.P384Point) {
	result := runar.P384Add(a, b)
	runar.Assert(runar.P384OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}

func (c *P384Primitives) VerifyMulGen(k runar.Bigint) {
	result := runar.P384MulGen(k)
	runar.Assert(runar.P384OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}
